docs/design.md

```diff
@@ -0,0 +1,111 @@
+underlying functionality, but with a number of changes driven by:
+
+* Trying to make it feel as "ruby native" as possible.
+* The ability to design an API from scratch that lacks many of the
+  constraints imposed both by the earlier much more limited functionality
+  of Hypothesis and the specifics of Python decorators and test
+  frameworks.
+
+## Differences
+
+The most fundamental API differences  between Hypothesis
+for Python and Hypothesis for Ruby are:
+
+* In Python we do a whole giant song and dance about exposing
+  functions for the test runner to call, while in Ruby we just
+  have a function which repeatedly calls a block and then fails.
+* in Python you specify a bunch of given parameters up front,
+  and then if you want values inline in the test you [explicitly
+  opt in to it](https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests),
+  while in Ruby this is not only the default but the only way to
+  get those values.
+* Strategies are called providers because strategy is a terrible
+  name that was originally intended to be internal and then leaked
+  into the public API because I wasn't thinking hard about naming.
+
+So for example:
+
+```ruby
+RSPec.describe "integer addition" do
+  it "commutes" do
+    hypothesis do
+      m = given integers
+      n = given integers
+      expect(m + n).to eq(n + m)
+    end
+  end
+end
+```
+
+```python
+@given(integers(), integers())
+def test_integers_commute(m, n):
+  assert m + n == n + m
+```
+
+The in-line style is slightly more verbose, but vastly more flexible
+and (I think) reads better. Also mixing in-line and up-front
+styles looks weird, and if we're going to have just one then
+the in-line approach is a strict superset of the functionality
+of the other.
+
+The main reason for these differences are:
+
+* Ruby blocks (and their relation to testing) make this approach
+  much more natural.
+* This functionality was not actually possible when the Hypothesis
+  for Python API was originally designed, which informed the way
+  its API looks.
+
+## Deliberate omissions
+
+The following are currently *not* part of the intended feature set
+of Hypothesis for Ruby:
+
+* Calls to `hypothesis` may not be nested.
+* None of [the functionality for adding explicit failures](https://hypothesis.readthedocs.io/en/latest/reproducing.html)
+  (but RSpec prints the random seed for your test run, and that will
+  be respected, so the seeding aspect of this happens for free).
+* There will be no equivalent to the [stateful testing](https://hypothesis.readthedocs.io/en/latest/stateful.html)
+  (but the very interactive nature of tests in the Ruby API means that
+  the generic state machine stuff is just something you can write in
+  your normal tests).
+* Testing will not be coverage guided (to be fair, it's barely coverage
+  guided in the Python version right now...)
+* Only one failure will be reported per test (though possibly behind
```

Comment:
Multiple failures will be quite nice-to-have in the future, but as with the above items not a requirement today.

Reply:
So I originally added this to the deliberate omissions because I thought it would be hard to integrate on the test framework side rather than due to any fundamental difficulties (this will be a required feature for the milestone around getting some of hypothesis-python parts working on this anyway!). On further investigation I *think* it's actually not too bad with RSpec using [aggregate_failures](http://rspec.info/documentation/3.5/rspec-expectations/RSpec/Matchers.html#aggregate_failures-instance_method). I will probably time box say half a day on getting this working in RSpec and if I get a viable prototype in that time then it'll be included, if not then it will be for a future version.

Reply:
In fact, even for minitest we can just depend on `rspec-expectations` and nothing bad will happen (this is an explicitly supported combination). So I think I'm going to remove this from the "won't be included" and add a "might not be included, time depending" section for this and `@example`.